src/enumerable.cr

```diff
@@ -637,6 +637,24 @@ module Enumerable(T)
     memo
   end
 
+  # Similar to `reduce`, but instead of raising when the input is empty,
+  # return `nil`
+  #
+  # ```
+  # ([] of Int32).reduce? { |acc, i| acc + i } # => nil
+  # ```
+  def reduce?
```

Comment:
If you take a look at `max_by` and `max_by?`, they use a `max_by_internal` implementation which returns a tuple: the value (here `memo`) and whether an element was found (`found`). Then `max_by` raises if `!found` while `max_by?` returns `nil` in that case.

I think we should do the same here, have a `reduce_internal` with the same approach to avoid duplicating this bit of implementation logic.

Reply:
Happy to refactor - this guidance would have been really helpful a couple of years ago when I embarked on this. There's someone wanting the functionality now (per the previous PR) - so I can try and get to this in a few days/weeks. The PR is editable by the maintainers if you wanted to improve it.

Reply:
We can always merge this and refactor later :-)

Thank you!